pkg/clients/nodepool/nodepool.go

```diff
@@ -0,0 +1,364 @@
+		}
+	}
+
+	spec.Version = gcp.LateInitializeString(spec.Version, in.Version)
+}
+
+// newAutoscalingUpdateFn returns a function that updates the Autoscaling of a node pool.
+func newAutoscalingUpdateFn(in *v1alpha1.NodePoolAutoscaling) UpdateFn {
+	return func(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+		out := &container.NodePool{}
+		GenerateAutoscaling(in, out)
+		update := &container.SetNodePoolAutoscalingRequest{
+			Autoscaling: out.Autoscaling,
+		}
+		return s.Projects.Locations.Clusters.NodePools.SetAutoscaling(name, update).Context(ctx).Do()
+	}
+}
+
+// newManagementUpdateFn returns a function that updates the Management of a node pool.
+func newManagementUpdateFn(in *v1alpha1.NodeManagementSpec) UpdateFn {
+	return func(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+		out := &container.NodePool{}
+		GenerateManagement(in, out)
+		update := &container.SetNodePoolManagementRequest{
+			Management: out.Management,
+		}
+		return s.Projects.Locations.Clusters.NodePools.SetManagement(name, update).Context(ctx).Do()
+	}
+}
+
+// newGeneralUpdateFn returns a function that updates a node pool.
+func newGeneralUpdateFn(in *v1alpha1.NodePoolParameters) UpdateFn {
+	return func(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+		return s.Projects.Locations.Clusters.NodePools.Update(name, GenerateNodePoolUpdate(in)).Context(ctx).Do()
+	}
+}
+
+func noOpUpdate(ctx context.Context, s *container.Service, name string) (*container.Operation, error) {
+	return nil, nil
+}
+
+// UpdateFn returns a function that updates a node pool.
+type UpdateFn func(context.Context, *container.Service, string) (*container.Operation, error)
+
+// IsUpToDate checks whether current state is up-to-date compared to the given
+// set of parameters.
+func IsUpToDate(in *v1alpha1.NodePoolParameters, currentState container.NodePool) (bool, UpdateFn) {
+	currentParams := &v1alpha1.NodePoolParameters{}
+	LateInitializeSpec(currentParams, currentState)
+	if !cmp.Equal(in.Autoscaling, currentParams.Autoscaling) {
+		return false, newAutoscalingUpdateFn(in.Autoscaling)
+	}
+	if !cmp.Equal(in.Management, currentParams.Management) {
+		return false, newManagementUpdateFn(in.Management)
+	}
+	// Ignore references, Cluster and InitialNodeCount because they are not
+	// reflected in the container.NodePool object.
+	if !cmp.Equal(in, currentParams,
+		cmpopts.IgnoreInterfaces(struct{ resource.AttributeReferencer }{}),
+		cmpopts.IgnoreFields(v1alpha1.NodePoolParameters{}, "Cluster", "InitialNodeCount")) {
+		return false, newGeneralUpdateFn(in)
+	}
+	return true, noOpUpdate
+}
+
+// GetFullyQualifiedName builds the fully qualified name of the cluster.
+func GetFullyQualifiedName(p v1alpha1.NodePoolParameters, name string) string {
+	return strings.Replace(fmt.Sprintf(NodePoolNameFormat, p.Cluster, name), "/zones/", "/locations/", -1)
```

Comment:
@muvaf just made this update as part of testing. If using referencing a `zonal` cluster the path has `zones` where `locations` should be used. Have confirmed that this fix allows for referencing of both regional and zonal clusters.

Reply:
Basically whatever GKE gives you as `selfLink` cannot be put there right away, we have to change `zonal` to `locations` in all cases, right? That's weird.

Reply:
Right, or else we would have to use `Projects.Locations.Clusters.NodePools` sometimes and `Projects.Zones.Clusters.NodePools` other times. I prefer to just modify the path.